src/main/scala/com/wavesplatform/state2/BlockchainUpdaterImpl.scala

```diff
@@ -76,112 +72,153 @@ class BlockchainUpdaterImpl private(persisted: StateWriter with StateReader,
       bottomMemoryDiff.set(topMemoryDiff())
       topMemoryDiff.set(BlockDiff.empty)
     }
-    (liquidBlockCandidatesDiff().get(block.reference) match {
-      case Some(referencedLiquidDiff) =>
-        val asFirmBlock = referencedLiquidDiff.copy(heightDiff = 1)
-        ngHistoryWriter.appendBlock(block)(BlockDiffer.fromBlock(settings,
-          composite(currentPersistedBlocksState, () => asFirmBlock),
-          ngHistoryWriter.forgeBlock(block.reference).map(_._1), block)).map { case ((newBlockDiff, discraded)) =>
-          topMemoryDiff.set(Monoid.combine(topMemoryDiff(), asFirmBlock))
-          liquidBlockCandidatesDiff.set(Map(block.uniqueId -> newBlockDiff))
-          discraded
-        }
+    (ngState() match {
       case None =>
-        ngHistoryWriter.appendBlock(block)(BlockDiffer.fromBlock(
-          settings, currentPersistedBlocksState, ngHistoryWriter.lastBlock, block)).map { case ((newBlockDiff, discraded)) =>
-          liquidBlockCandidatesDiff.set(Map(block.uniqueId -> newBlockDiff))
-          discraded
+        lazy val logDetails = s"The referenced block(${block.reference})" +
+          s" ${if (historyWriter.contains(block.reference)) "exits, it's not last persisted" else "doesn't exist"}"
+        historyWriter.lastBlock match {
+          case Some(lastInner) if lastInner.uniqueId != block.reference =>
+            Left(BlockAppendError(s"References incorrect or non-existing block: " + logDetails, block))
+          case _ =>
+            historyWriter.lastBlock.map(_.uniqueId) match {
+              case Some(lastId) if lastId != block.reference => Left(GenericError(s"Parent ${block.reference} of new liquid block ${block.uniqueId}" +
+                s" does not match last block $lastId"))
+              case _ => BlockDiffer.fromBlock(settings, currentPersistedBlocksState, historyWriter.lastBlock, block).map((_, Seq.empty[Transaction]))
+            }
+        }
+      case Some(ng) if !ng.contains(block.reference) =>
+        Left(BlockAppendError(s"References incorrect or non-existing block", block))
+      case Some(ng) =>
+        val referencedLiquidDiff = ng.diffs.get(block.reference).get
```

Comment:
You can use just `ng.diff(block.reference)` here. But what if there is no such element? We'll get an exception and it will be forwarded to upper levels. Is that correct?

Reply:
Actually we won't, because `.diffs` and `.micros` are in sync, if we didn't go to ` Some(ng) if !ng.contains(block.reference)`, then both values of `ng.diffs` and next line of `ng.forgeBlock(block.reference).get` would be safe.

Reply:
Will fix `.get(..).get`